lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/address/ContainerBasedLocalAddressProvider.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.wake.remote.address;
+
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Tang;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.UnknownHostException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A LocalAddressProvider that uses <code>Inet4Address.getLocalHost().getHostAddress()</code>.
+ */
+public final class ContainerBasedLocalAddressProvider implements LocalAddressProvider {
+
+  private static final Logger LOG = Logger.getLogger(ContainerBasedLocalAddressProvider.class.getName());
+
+  private String cached = null;
+
+  /**
+   * The constructor is for Tang only.
+   */
+  @Inject
+  private ContainerBasedLocalAddressProvider() {
+    LOG.log(Level.FINE, "Instantiating ContainerBasedLocalAddressProvider");
+  }
+
+  @Override
+  public synchronized String getLocalAddress() {
+    if (null == cached) {
```

Comment:
this logic here is a bit hard to read, can we simplify it to:
'''
if(ipAddress == null) { throw new RuntimeException(message)}
...
if(!ipAddressFile.exists()||!ipAddressFile.isFile()) {throw new RuntimeException()}
...
```

Reply:
This just how the value is cached upon first read and reused later whenever needed. The code is similar to HostBasedLocalAddressProvider. So I left the variables ('cached') as is so it has parity with the other implementations of LocalAddressProvider interface.

---
In reply to: [196583917](https://github.com/sharathmalladi/reef/pull/5#discussion_r196583917) [](ancestors = 196583917)

Reply:
I mean to remove nested 'if' in below sections, by changing if(a!=b) {} else{throw} to if(a==b) throw;

---
In reply to: [196968145](https://github.com/sharathmalladi/reef/pull/5#discussion_r196968145) [](ancestors = 196968145,196583917)

Reply:
Discussed with Chenxi. Will flatten the ifs.

---
In reply to: [196971818](https://github.com/sharathmalladi/reef/pull/5#discussion_r196971818) [](ancestors = 196971818,196971584,196968145,196583917)